### PR TITLE
Return a Template kind in v1beta3 from process

### PR DESCRIPTION
**Pull request by smarterclayton**
Will allow us to expose more info to clients.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@bparees review

File: test/integration/template_test.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		// keep existing values
+		if obj.Objects[0].(*runtime.Unstructured).Object["portalIP"] != "1.2.3.4" {
+			t.Fatalf("unexpected object: %#v", obj)
+		}
+		// replace a value
+		if obj.Objects[0].(*runtime.Unstructured).Object["id"] != "test-tester" {
+			t.Fatalf("unexpected object: %#v", obj)
+		}
+		// clear namespace
+		if obj.Objects[0].(*runtime.Unstructured).Object["namespace"] != "" {
```
Comment (bparees) [inline test/integration/template_test.go]: why do we expect namespace to be cleared here?

**Comment by bparees** [comment]
so this breaks the "osc process -f file.json | osc create -f -" flow, right?

File: test/integration/template_test.go (hunk 1 above)
Comment (smarterclayton) [inline test/integration/template_test.go]: We've always cleared the namespace.  Namespace is imposed by create, not by template.

----- Original Message -----

**Comment by smarterclayton** [comment]
It does, I have to make process return list.  Good catch (and bad for the fact we don't have tests for it).  I need to add a v1beta3 loop for hack/test-cmd

----- Original Message -----

**Comment by bparees** [comment]
Doesn't e2e do a process | create flow?

**Comment by smarterclayton** [comment]
Fixed

File: pkg/template/registry/rest.go [hunk 2]
```diff
@@ -120,5 +120,5 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 	if s.legacyReturn {
 		return cfg, nil
 	}
-	return &kapi.List{Items: cfg.Items}, nil
+	return tpl, nil
```
Comment (bparees) [inline pkg/template/registry/rest.go]: haven't you lost the label application aspect of template processing here?

**Comment by bparees** [comment]
this is still going to break the create from template web console flow, isn't it?  how are we allowed to change the return type of this api?  is there a v1beta1 conversion somewhere that turns a Template into a List?

also @jcantrill is working on create from template for eclipse, so he should probably be aware of this coming change, if i understand it correctly.

File: pkg/template/registry/rest.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/template/registry/rest.go]: If I did.... we didn't have any test for it.

----- Original Message -----

**Comment by smarterclayton** [comment]
Only in v1beta3

----- Original Message -----

File: pkg/template/registry/rest.go (hunk 2 above)
Comment (bparees) [inline pkg/template/registry/rest.go]: talk to @csrwng or @mfojtik i think one of them added this feature.

**Comment by bparees** [comment]
@smarterclayton i'm missing something..what logic is ensuring that the v1beta1 endpoint still returns a List?

File: pkg/template/registry/rest.go (hunk 2 above)
Comment (csrwng) [inline pkg/template/registry/rest.go]: https://github.com/openshift/origin/blob/master/pkg/template/registry/rest_test.go#L61

Comment (bparees) [inline pkg/template/registry/rest.go]: correct me if i'm wrong but that test should be failing spectacularly (type cast error) with this change.

Comment (smarterclayton) [inline pkg/template/registry/rest.go]: Conversion handles it.

----- Original Message -----

Comment (smarterclayton) [inline pkg/template/registry/rest.go]: Fix label application (moved it into Process where it belongs) and added tests (so we actually, you know, know that it works).

----- Original Message -----

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
The legacy flag ensures that in rest.go

**Comment by bparees** [comment]
1) legacy makes it return a Config, not a List.....
2) what conversion?  it's not in this PR... i don't see anything that's converting Lists or Configs to Templates or vice versa
3) not sure why @csrwng 's existing tests weren't failing w/ this?  because they were using the v1b1 api?

**Comment by smarterclayton** [comment]
----- Original Message -----
> 1) legacy makes it return a Config, not a List.....
> 2) what conversion?  it's not in this PR... i don't see anything that's
> converting Lists or Configs to Templates or vice versa

Configs and Lists are converted to each other.

> 3) not sure why @csrwng 's existing tests weren't failing w/ this?  because
> they were using the v1b1 api?

Not sure.  Didn't look.

**Comment by smarterclayton** [comment]
If you have more comments give em [merge]

**Pull request merged**